h2o-mapreduce-generic/src/main/java/water/hadoop/h2odriver.java

```diff
@@ -528,36 +524,54 @@ else if (type == MapperToDriverMessage.TYPE_EXIT) {
    */
   class CallbackManager extends Thread {
     private ServerSocket _ss;
+    private int _numNodes;
 
     // Nodes and socks
-    private final HashSet<String> _dupChecker = new HashSet<String>();
-    private final ArrayList<String> _nodes = new ArrayList<String>();
-    private final ArrayList<Socket> _socks = new ArrayList<Socket>();
+    private final HashMap<String, Integer> _dupChecker = new HashMap<>();
+    final ArrayList<String> _nodes = new ArrayList<>();
+    final ArrayList<Socket> _socks = new ArrayList<>();
 
-    public void setServerSocket (ServerSocket value) {
-      _ss = value;
+    CallbackManager(ServerSocket ss, int numNodes) {
+      _ss = ss;
+      _numNodes = numNodes;
     }
 
-    public void registerNode (String ip, int port, Socket s) {
+    public void registerNode (String ip, int port, int attempt, Socket s) {
       synchronized (_dupChecker) {
         String entry = ip + ":" + port;
 
-        if (_dupChecker.contains(entry)) {
-          // This is bad.
-          System.out.println("ERROR: Duplicate node registered (" + entry + "), exiting");
-          System.exit(1);
+        if (_dupChecker.containsKey(entry)) {
+          int prevAttempt = _dupChecker.get(entry);
+          if (prevAttempt == attempt) {
+            // This is bad.
+            fatalError("Duplicate node registered (" + entry + "), exiting");
+          } else if (prevAttempt > attempt) {
+            // Suspicious, we are receiving attempts out-of-order, stick with the . 
```

Comment:
small note: missing end of the sentence? Stick with the .. ?

Reply:
yep! good catch, thank you